DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/task_sequence_hardware_reuse/task_sequence/src/main.cpp

```diff
@@ -0,0 +1,134 @@
+#include <iostream>
+
+// oneAPI headers
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/ext/intel/experimental/task_sequence.hpp>
+#include <sycl/sycl.hpp>
+
+#include "exception_handler.hpp"
+
+// Forward declare the kernel name in the global scope. This is an FPGA best
+// practice that reduces name mangling in the optimization reports.
+class VectorOpID;
+
+constexpr int kVectSize = 3;
+constexpr int N = 5;
+struct D3Vector {
+  float d[3];
+};
+
+float OpSqrt(D3Vector val, const D3Vector coef) {
+  float res = sqrt(val.d[0] * coef.d[0] + val.d[1] * coef.d[1] + val.d[2] * coef.d[2]);
+  return res;
+}
+
+struct VectorOp {
+  D3Vector *a_in;
+  float *z_out;
+  int len;
+
+  void operator()() const {
+    constexpr D3Vector coef1 = {0.2, 0.3, 0.4};
+    constexpr D3Vector coef2 = {0.6, 0.7, 0.8};
+
+    D3Vector new_item;
+    
+    // Object declarations of a parameterized task_sequence class must be local, which means global declarations and dynamic allocations are not allowed.
+    // Declare the task sequence object outside the for loop so that the hardware can be shared at the return point.
+    sycl::ext::intel::experimental::task_sequence<OpSqrt> task_a;
+    
+    for (int i =0; i < len; i++) {
+      task_a.async(a_in[i], coef1);
+    }
+
+    for (int i =0; i < len; i++) {
+      new_item.d[i] = task_a.get();
+    }
```

Comment:
Did the compiler fuse these loops? Check to see if there are still two loops in the report or just one. There is a bug right now where the compiler doesn't call attention to the fact that it fused loops:
https://hsdes.intel.com/appstore/article/#/14017424650 

I think you can avoid the issue entirely by calling `async()` and `get()` in the same loop.

```suggestion
    for (int i =0; i < len; i++) {
      task_a.async(a_in[i], coef1);
      new_item.d[i] = task_a.get();
    }
```

If you have questions about `task_sequence` while I'm out next week, you can reach out to Adel Ejjeh (adel.ejjeh@intel.com)

Reply:
@whitepau I can still see 2 loops in the report. I am not changing the code at the moment. Let me know if you think it is necessary.

You may view the report here: https://psg-sc-arc.sc.intel.com/nfs/site/disks/psg_data_49/choonnee/psg_zsc7/C++SYCL_FPGA/Tutorials/Features/experimental_240408/task_seq_hw_reuse/tbuild_a7/task_sequences.fpga_sim.prj/reports/report.html#

Reply:
so here's the thing. 

This code is fine, **so long as** the number of loop iterations is _less than_ the latency of the task function of `task_a`. In this case, the report shows a latency of 73 cycles, so you're ok.

If the the number of loop iterations is larger, then the design will hang since the task sequence will backpressure after its data path is full.

There are two options here (i think both are valid, so you can pick):

1. This would be an excellent place to introduce task capacity; you could add a part 4 that uses a much smaller task function with a smaller latency, and a probably a larger number of iterations.

2. fuse the loops in this sample, and make a separate tutorial on task sequence pipelining.